### DISCUSSION TITLE
Possible bug in FLTK?

### DISCUSSION THREAD

**Discussion by 4JX** [fltk-rs/fltk-rs#1101] [General]
I'm encountering some strange behavior in my FLTK UI. One of my `IntInput`s keeps getting set to 0 when I lift the keys I'm pressing on the keyboard (which are used for a shortcut). See: https://ibb.co/z7cnSw6.

It only occurs when the keyboard is used. Manually clicking or forcing the code to run do not appear to have the issue.

For it to trigger, one of the "Presets" which are cycled through needs to have an effect that [runs in a loop](https://github.com/4JX/L5P-Keyboard-RGB/blob/7dfd8f541c276c841c16d18d3c4cf4daac787501/src/keyboard_manager.rs#L77-L87) that is exited with the use of conditional breaks and `AtomicBool`s (as this code runs on a separate thread to avoid blocking the main one).

Only once the loop effect has been activated the issue will start to manifest, otherwise the app will run just fine.

The issue does go away if I comment out [this line](https://github.com/4JX/L5P-Keyboard-RGB/blob/7dfd8f541c276c841c16d18d3c4cf4daac787501/src/gui/app.rs#L345), which is in charge of activating/deactivating all the inputs based on the selected effect.

**Comment by MoAlyousef** [comment]
I think it's the handle method where you handle KeyUp and you set the input value to 0 on error. Normally you would want to handle app::event_key() in KeyUp. Here I don't really see a reason to add a handler for it.

**Reply by 4JX** [reply]
> Here I don't really see a reason to add a handler for it.

The app needs to tell a separate thread to update each time the color values change, that is, when the user types a number in the `IntInput`s. To do that it needs to signal the other thread to break out of its loop (should it be in one) (used for animated effects by spamming "change rgb values" to the keyboard), hence this bit of code:

```
stop_signals.store_true();
tx.send(Message::Refresh).unwrap();
```

After some more investigation the issue seems to reside in the fact that after disabling and re-enabling the tiles, FLTK chooses to arbitrarily put the cursor (black line that indicates I am typing in that slot) on the bottom red input, which causes [this to execute](https://github.com/4JX/L5P-Keyboard-RGB/blob/dev/src/gui/color_tiles.rs#L343-L346) (from [this](https://github.com/4JX/L5P-Keyboard-RGB/blob/92ae3e3a12275416cd7a439fb90ae16d3c1d6223/src/gui/color_tiles.rs#L246) callback) setting all the "R:" inputs to the value of the bottom one, in this case being 0.

`Enums::Keys` does not contain any reference to number keys, so I'd guess I would either need to check if the value has changed (which would warrant an update of the inputs) or use another library like `device_query` to check for what key is being pressed.

**Reply by MoAlyousef** [reply]
You can construct keys from a const function Key::from_char('1'), or match by converting the Key to char like here:
https://github.com/fltk-rs/fltk-egui/blob/e1e21f1586c859d8d65e2b0c8238ab5d881b1560/src/lib.rs#L318

Another option is to use input.set_trigger(CallbackTrigger::Changed), and add the logic in the set_callback method.

**Reply by 4JX** [reply]
The latter did the job :+1:  https://github.com/4JX/L5P-Keyboard-RGB/commit/8bd418d04ec204cc15ce93c5eed5b08b190dcef6
